src/napari/_vispy/utils/text.py

```diff
@@ -71,3 +73,171 @@ def _has_visible_text(layer: Points | Shapes) -> bool:
     ):
         return False
     return len(layer._indices_view) != 0
+
+
+def get_text_width_height(text: Text) -> tuple[float, float]:
+    """Get the screen space width and height of a vispy text visual."""
+    if isinstance(text.text, str):
+        strings = [text.text]
+    elif isinstance(text.text, list):
+        strings = text.text
+    else:
+        raise TypeError('Text should either be a string or a list of strings')
+
+    top_left_corners = []
+    bottom_right_corners = []
+    for string in strings:
+        if string == '':
+            continue
+
+        # this is a private vispy function that calculates the Vertex Buffer Object
+        # for the text to be sent to the shaders. This normallly only happens on
+        # the fly when rendering, but we need it here because it allows us to
+        # calculate the bounding box of the text as it would be when rendered
+        buffer = _text_to_vbo(
+            string, text._font, *text._anchors, text._font._lowres_size
+        )  # type: ignore
+
+        pos = buffer['a_position']
+        top_left_corners.append(pos.min(axis=0))
+        bottom_right_corners.append(pos.max(axis=0))
+
+    top_left = np.min(top_left_corners, axis=0) if top_left_corners else (0, 0)
+    bottom_right = (
+        np.max(bottom_right_corners, axis=0)
+        if bottom_right_corners
+        else (0, 0)
+    )
+
+    font_size = get_text_font_size(text)
+
+    # these magic numbers (1.2 and 1.3) are from trial and error
+    return (bottom_right[0] - top_left[0]) * font_size * 1.3, (
+        bottom_right[1] - top_left[1]
+    ) * font_size * 1.2
+
+
+def get_text_font_size(text: Text) -> float:
+    """Get the logical font size of a text visual, rescaled by dpi."""
+    # use 96 as the vispy reference dpi for historical reasons
+    dpi_scale_factor = 96 / text.transforms.dpi if text.transforms.dpi else 1
+
+    return text.font_size * dpi_scale_factor
+
+
+# vendored from vispy/visuals/text/text.py, but removing all lines referring to context flushing,
+# canvas and viewport which cause issues. We only need to calculate sizes anyways.
+@no_type_check
```

Comment:
Why not type it correctly to help other devs to understand it?

Reply:
Then could port typing back to vispy for this?

Reply:
I think I'd rather rewrite it from scratch, which is already a pain :sweat_smile: Unless you mean just type the arguments and return, in which case it's probably reasonably doable.